language/move-prover/tests/sources/functional/let.move

```diff
@@ -0,0 +1,71 @@
+module TestLet {
+
+    spec module {
+        pragma verify = true;
+    }
+
+    resource struct R {
+        x: u64
+    }
+
+    // Spec Block Let
+    // ==============
+
+    fun spec_let_with_result(a: u64, b: u64): (u64, u64) {
+        (a + 1 + b, a + b)
+    }
+    spec fun spec_let_with_result {
+        let x = a + 1;
```

Comment:
The semantics appears to be macro definition and substitution.  I am wondering if the use of word "let" for this feature will lead to confusion.  As you explained in the PR description, perhaps the right way to think of this feature as a let is that we are let-binding a lambda which is evaluated later when used in a function.  Is there a way to somehow indicate this in the syntax?

Reply:
Perhaps use some keyword other than let, e.g., let-fun?

Reply:
Yet another option might be to use let keyword but require the lambda notation always  even if there are no explicit parameters, e.g.,

let y = || x + b;

Reply:
It's actually not substitution but polymorphic let (in Hindley-Milner sense -- ML language). Yes, they are lazily evaluated, but they are not substituted but replaced by functions which are called as needed. I feel since the prover doesn't really makes any difference between "strict" and "lazy" evaluation, we do not need to complicate the mental model by additional keywords and syntax. I agree we should document this lazy aspect once we write the language definition, just so people aren't shy of defining lots of lets even if they are only used inside of conditional expressions. 

For the "let-fun" and the "let y = ||...", I would argue similar.